DOCS-11871-batchTuning-v44-duke (#2107)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/tuning-batch-processing.adoc

<batch:job jobName="test-batch" blockSize="${batch.block.size}">
----

== Max Concurrency

The maximum number of threads for processing a Batch scope defaults to 2 times the number of cores that the JVM detects. On fractional core deployments, such as CloudHub and Anypoint Runtime Fabric, the JVM detects a _non-fractional_ number of cores. You can manually increase or lower the automated concurrency setting. The Max Concurrency setting enables you to set an upper limit on the number of threads to use for a batch job instance that, for example, has insufficient records to warrant more threads.

> The follow example overrides the automated `maxConcurrency` default to a setting that is loaded from a property:

Reviewer: ```suggestion
The following example overrides the automated `maxConcurrency` default to a setting that is loaded from a property:
```

Reviewer: nonfractional, per style ref.

[source,xml,linenums]
----
<batch:job jobName="an-example-batch" maxConcurrency="${batch.block.maxConcurrency}">
----

== Batch Aggregators

> One million round trips across a network, whether on single or multiple threads, impacts performance. Communication with SaaS systems can easily consume the quota of API calls. Instead of processing records one by one, you can improve performance by using the Batch Aggregator scope to perform bulk operations on groups of records. The scope uses an Aggregator Size setting to divide the records into groups. When determining the number of records to process in bulk, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.

Reviewer: I think that the first sentence might be more powerful as...:
"One million round trips across a network impacts performance, whether on...."

Reviewer: Consider breaking this para into two. It seems like a lot of info for one para, and I think you might be able to break between the point that bulk ops help performance, and how to determine the optimum number of bulk ops.

Reviewer: Do we need to mention that Batch scopes include Batch Step scopes (Line 37) and Batch Job scopes (Line 53)?

Author: ```suggestion
One million round trips across a network impacts performance, whether on single or multiple threads. Communication with SaaS systems can easily consume the quota of API calls. Instead of processing records one by one, you can improve performance by using the Batch Aggregator scope to perform bulk operations on groups of records. The scope uses an Aggregator Size setting to divide the records into groups. 

When determining the number of records to process in bulk, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.
```

The following example shows a Batch Step scope with a Batch Aggregator scope that is set to a fixed size of `50`:

[source,xml,linenums]
----
<batch:step name="Batch_Step_PushToXYZ" >
  <batch:aggregator doc:name="Batch Aggregator" size="50">
     <!-- code for an array of 50 (or fewer) records -->
  </batch:aggregator>
</batch:step>
----

> If the aggregator contains an operation that streams data, you can activate the scope's Streaming option  (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).

Reviewer: ```suggestion
If the aggregator contains an operation that streams data, you can activate the scope's Streaming option (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).
```

== Multiple Concurrent Instances of a Batch Job Scope

> Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy, each of which is tracked separately by its batch instance ID: 

Author: ```suggestion
Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set one of the scheduling strategies, each of which is tracked separately by its batch instance ID: 
```

> Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set one of the scheduling strategies, each of which is tracked separately by its batch instance ID: 

Reviewer: ```suggestion
Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set one of the following scheduling strategies, each of which is tracked separately by its batch instance ID: 
```

Reviewer: The following two items s/b bullets, right?

> `ORDERED_SEQUENTIAL`: Default scheduling strategy that limits the impact of concurrent instances by processing the instances in the order they are triggered. This strategy is useful when non-sequential processing of the same batch job can cause issues, such as data inconsistency, or reach resource limitations. 

Reviewer: nonsequential, per style ref.

Author: ```suggestion
* `ORDERED_SEQUENTIAL`: Default scheduling strategy that limits the impact of concurrent instances by processing the instances in the order they are triggered. This strategy is useful when nonsequential processing of the same batch job can cause issues, such as data inconsistency, or reach resource limitations. 
```

> `ROUND_ROBIN`: Scheduling strategy that shares the threads with batch job instances that are "in flight". 

Author: ```suggestion
* `ROUND_ROBIN`: Scheduling strategy that shares the threads with batch job instances that are "in flight". 
```

<batch:job jobName="test-batch" blockSize="${batch.block.size}">
----

== Max Concurrency

> The maximum number of threads for processing a Batch scope defaults to 2 times the number of cores that the JVM detects. On fractional core deployments, such as CloudHub and Anypoint Runtime Fabric, the JVM detects a _non-fractional_ number of cores. You can manually increase or lower the automated concurrency setting. The Max Concurrency setting enables you to set an upper limit on the number of threads to use for a batch job instance that, for example, has insufficient records to warrant more threads.

Author: ```suggestion
The maximum number of threads used for processing within a Batch scope defaults to 2 times the number of cores that the JVM detects. On fractional core deployments, such as CloudHub and Anypoint Runtime Fabric, the JVM detects a _non-fractional_ number of cores. You can manually increase or lower the automated concurrency setting. The Max Concurrency setting enables you to set an upper limit on the number of threads to use for a batch job instance that, for example, has insufficient records to warrant more threads.
```

The following example overrides the automated `maxConcurrency` default to a setting that is loaded from a property:

[source,xml,linenums]
----
<batch:job jobName="an-example-batch" maxConcurrency="${batch.block.maxConcurrency}">
----

== Batch Aggregators

> One million round trips across a network impacts performance, whether on single or multiple threads. Communication with SaaS systems can easily consume the quota of API calls. Instead of processing records one by one, you can improve performance by using the Batch Aggregator scope to perform bulk operations on arrays of records. The scope uses an Aggregator Size setting to limit the number of records in each array. 

Author: @nathan-at-mulesoft  Or does it make more sense to say "an array of records" (that is, one array vs. multiple) and "number of records in the array" like this? 

"Instead of processing records one by one, you can improve performance by using the Batch Aggregator scope to perform bulk operations on _an array_ of records. The scope uses an Aggregator Size setting to limit the number of records in _the_ array."

> When determining the number of records to process in bulk, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.

Author: ```suggestion
When determining the number of records to process in a group, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.
```

The following example shows a Batch Step scope with a Batch Aggregator scope that is set to a fixed size of `50`:

[source,xml,linenums]
----
<batch:step name="Batch_Step_PushToXYZ" >
  <batch:aggregator doc:name="Batch Aggregator" size="50">
     <!-- code for an array of 50 (or fewer) records -->
  </batch:aggregator>
</batch:step>
----

If the aggregator contains an operation that streams data, you can activate the scope's Streaming option (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).

> == Multiple Concurrent Instances of a Batch Job Scope

Author: ```suggestion
== Multiple Concurrent Instances of a Batch Job
```
